src/core/SampleBuffer.cpp

```diff
@@ -1017,30 +1025,75 @@ void SampleBuffer::visualize(
 	const float ySpace = h * 0.5f;
 	const int nbFrames = focusOnRange ? toFrame - fromFrame : m_frames;
 
-	const int fpp = qBound<int>(1, nbFrames / w, 20);
-	QPointF * l = new QPointF[nbFrames / fpp + 1];
-	QPointF * r = new QPointF[nbFrames / fpp + 1];
-	int n = 0;
+	const int fpp = std::max(1, nbFrames / w);
+	// There are 2 possibilities: Either nbFrames is bigger than
+	// the width, so we will have width * 2 points, or nbFrames is
+	// smaller than the width (fpp = 1) and we will have nbFrames * 2
+	// points
+	const int totalPoints = nbFrames > w
+		? w * 2
+		: nbFrames * 2;
+	std::vector<QPointF> fMax(totalPoints);
+	std::vector<QPointF> fRms(totalPoints);
+	int curPixel = 0;
 	const int xb = dr.x();
 	const int first = focusOnRange ? fromFrame : 0;
 	const int last = focusOnRange ? toFrame - 1 : m_frames - 1;
+	// When the number of frames isn't perfectly divisible by the
+	// width, the remaining frames don't fit the last pixel and are
+	// past the visible area. lastVisibleFrame is the index number of
+	// the last visible frame.
+	const int visibleFrames = (fpp * w);
+	const int lastVisibleFrame = focusOnRange
+		? fromFrame + visibleFrames - 1
+		: visibleFrames - 1;
 
-	for (int frame = first; frame <= last; frame += fpp)
+	for (int frame = first; frame <= last && frame <= lastVisibleFrame; frame += fpp)
 	{
-		auto x = xb + ((frame - first) * double(w) / nbFrames);
+		float maxData = -1;
+		float minData = 1;
+
+		float rmsData[2] = {0, 0};
+
+		// Find maximum and minimum samples within range
+		for (int i = 0; i < fpp && frame + i <= last; ++i)
+		{
+			for (int j = 0; j < 2; ++j)
+			{
+				auto curData = m_data[frame + i][j];
+
+				if (curData > maxData) { maxData = curData; }
+				if (curData < minData) { minData = curData; }
+
+				rmsData[j] += curData * curData;
+			}
+		}
+
+		const float trueRmsData = (rmsData[0] + rmsData[1]) * 0.5 / fpp;
```

Comment:
Checking in godbolt, it looks like we can make this save some cycles by using `/ 2` instead of `* 0.5`.

Reply:
I'm not convinced by this. The number of cycles doesn't correspond exactly to the number of instructions - some are more expensive than others. While the division version has one less instruction in total, it uses a division instruction instead of a multiplication instruction, which is considerably more expensive. Besides, at optimisation levels `-O1` and above, identical code is generated for each (the compiler converts the division to a multiplication, because it's faster). Guessing at microoptimisations like this is nearly useless; actually measuring the execution time is the only sensible thing to do.

(That said, I prefer `/ 2` here anyway - it stands out more clearly to me as an average that way.)

Reply:
Yeah, I did say *might* save some cycles. 😁 You're right though it's not proper to try and optimize like that. I was curious and wanted to see what both turned into.